### PR TITLE
Attitude Orbit Simulations for PSim Integration

### PR DESCRIPTION
**Pull request by shihaocao**
# Attitude Orbit Simulations for PTest Integration

The main features added by this PR are:

* `DualAttitudeOrbitGnc` and `SingleAttitudeOrbitGnc` simulation types/models.
* Added a `HillFrame` model interface to support relative position and velocity calculations.
* Basic initial conditions for a deployment case (see `config/parameters/truth/deployment.txt`).
* Small update(s) to the attitude orbit model:
  * Lazy evaluation quaternion function.
  * Use body frame magnetic field instead of intertial frame.
* Other random bug fixes throughout PSim involving infinite recursion causing a segfault, a coordinate frame issue, and other miscelaneous issues.
* Initial conditions for deployment and detumble testcases.

Note for future debugging, it's possible to just run `gdb python` and the call the following from inside `gdb` to get a stack trace out of PSim on assertion/segfault issues:

    gdb python
    (gdb) r -m psim <arguments producing an error>

## Verification

We were able to successfully detumble in PSim standalone:

detumble

The detumbler for PSim standalone isn't included here and will be included with a seprate PR as it was super hacky.

@shihaocao can also confirm this version of PSim allowed us to run a successful detumble case in HOOTL!

### CODE REVIEW

File: config/parameters/detumble.txt [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+truth.leader.orbit.v  -7236.19   -1695.2    -1708.76
+
+# Follower spacecraft attitude and orbit initial conditions.
+
+truth.follower.orbit.r  -2.16425e6  4.60331e6  4.59251e6
+truth.follower.orbit.v  -7236.17   -1695.04   -1708.5
+
+truth.follower.attitude.q.body_eci -0.282557 -0.604902  -0.71832   -0.195632
+truth.follower.attitude.w          -0.247035  0.0574472  0.0659878
+truth.follower.wheels.w             0.0       0.0        0.0
```
Comment (kylekrol) [inline config/parameters/detumble.txt]: Initial conditions for a detumble case. These were generated form a deployment case using the snapshot feature with the Python command listed in the file's header.

File: config/parameters/truth/base.txt [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+truth.leader.wheels.J      135.0e-7
+truth.leader.wheels.w_max  677.0
+
+# Follower spacecraft attitude and orbit shared initial conditions.
+
+truth.follower.m  5.0
+truth.follower.J  0.03798 0.03957 0.00688
+
+truth.follower.wheels.J      135.0e-7
+truth.follower.wheels.w_max  677.0
```
Comment (kylekrol) [inline config/parameters/truth/base.txt:23]: Set of configuration constants shared across all simulations.

File: config/parameters/truth/deployment.txt [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+#
+# The velocity vector is slightly perturbed from the leader's to allow them to
+# drift apart.
+
+truth.follower.attitude.q.body_eci  0.0  0.0  0.0 1.0
+truth.follower.attitude.w           0.2 -0.15 0.1
+truth.follower.wheels.w             0.0  0.0  0.0
+
+truth.follower.orbit.r  6.8538e6 0.0      0.0
+truth.follower.orbit.v  0.0      5.3954e3 5.3951e3
```
Comment (kylekrol) [inline config/parameters/truth/deployment.txt:27]: Deployment initial conditions where the satellites start at the same location with a small relative velocity and different angular rates.

File: include/psim/truth/attitude_orbit.yml [hunk 4]
```diff
@@ -84,5 +88,5 @@ gets:
       type: Real
     - name: "{prefix}.earth.q.eci_ecef"
       type: Vector4
-    - name: "{prefix}.{satellite}.environment.b.eci"
+    - name: "{prefix}.{satellite}.environment.b.body"
```
Comment (kylekrol) [inline include/psim/truth/attitude_orbit.yml:91]: We needed magnetic field in the body frame not ECI - my bad originally.

File: include/psim/truth/attitude_orbit.yml [hunk 5]
```diff
@@ -76,6 +76,10 @@ adds:
       comment: >
           Magnetic moment commanded of the magnetic torque rods. This value is
           defaulted back to zero at each step.
+    - name: "{prefix}.{satellite}.attitude.L"
+      type: Lazy Vector3
+      comment: >
+          Angular momentum of the spacecraft in the body frame.
```
Comment (kylekrol) [inline include/psim/truth/attitude_orbit.yml:82]: Added a lazily evaluated angular momentum field for debugging.

File: python/psim/simulation.py [hunk 6]
```diff
@@ -53,6 +53,15 @@ def __setitem__(self, name, value):
...
 
+    def get(self, name):
+        """Attempts to retrieve a state field from the underlying simulation.
+        Upon failure, 'None' is returned.
+        """
+        try:
+            return self[name]
+        except RuntimeError:
+            return None
+
```
Comment (kylekrol) [inline python/psim/simulation.py:64]: Added a way to get a field without potentially raising an exception.

File: src/psim/truth/earth.cpp [hunk 7]
```diff
@@ -42,7 +42,7 @@ Vector4 EarthGnc::prefix_earth_q_ecef_eci() const {
 }
 
 Vector4 EarthGnc::prefix_earth_q_eci_ecef() const {
-  auto const &q_ecef_eci = this->Super::prefix_earth_q_eci_ecef.get();
+  auto const &q_ecef_eci = this->Super::prefix_earth_q_ecef_eci.get();
```
Comment (kylekrol) [inline src/psim/truth/earth.cpp:45]: This was the bug causing infinite recursion and a segfault!

File: src/psim/truth/transform_direction.cpp [hunk 8]
```diff
@@ -47,19 +47,19 @@ Vector3 TransformDirectionBody::vector_ecef() const {
...
 }
 
 Vector3 TransformDirectionEcef::vector_body() const {
   auto const &d_eci = Super::vector_eci.get();
-  auto const q_eci_body = prefix_satellite_attitude_q_eci_body->get();
+  auto const q_body_eci = prefix_satellite_attitude_q_body_eci->get();
 
   Vector3 d_body;
-  gnc::utl::rotate_frame(q_eci_body, d_eci, d_body);
+  gnc::utl::rotate_frame(q_body_eci, d_eci, d_body);
```
Comment (kylekrol) [inline src/psim/truth/transform_direction.cpp:62]: These were breaking the attitude estimator as they're incorrectly converting the magnetic field reading to the body frame.

File: config/parameters/truth/base.txt (hunk 2 above)
Comment (tanishqaggarwal) [reply config/parameters/truth/base.txt:23]: A lot of these constants look like they'd also be part of the existing C++ GNC code. Can we use those for configuration?

File: config/parameters/truth/detumble.txt [hunk 9]
```diff
@@ -0,0 +1,26 @@
+# Initial conditions for simulations starting during detumble. These initial
+# conditions were generated with the following command:
+#   python -m psim -s 10000 -c truth/base,truth/deployment --snapshot snapshot.txt DualAttitudeOrbitGnc
+#
+
+# Time model initial conditions.
+
+truth.t.ns  1700000000000
```
Comment (tanishqaggarwal) [inline config/parameters/truth/detumble.txt:8]: This parameter will change if the flight software cycle time changes (which is possible.) Tracked in https://github.com/pathfinder-for-autonomous-navigation/FlightSoftware/issues/532

File: python/psim/plugins/snapshot.py [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+        self._snapshot = args.snapshot
+        log.info('Saving simulation snapshot to "%s" upon simulation termination.', self._snapshot)
+
+    def cleanup(self, sim):
+        super(Snapshot, self).cleanup(sim)
+
+        if not self._snapshot:
+            return
+
+        _fields = [
```
Comment (tanishqaggarwal) [inline python/psim/plugins/snapshot.py:48]: Would there be any reason for a list like this to be passed in as, say, a filename via a psim command line option? Like do we expect this set of parameters to change? If so, might be worth putting in the work now to have it work?

File: python/psim/simulation.py (hunk 6 above)
Comment (tanishqaggarwal) [reply python/psim/simulation.py:64]: Is this for the PTest integration? Isn't it dangerous for stuff like this to fail silently?

File: src/psim/simulations/single_orbit.cpp [hunk 11]
```diff
@@ -22,7 +22,7 @@
 // SOFTWARE.
```
Comment (tanishqaggarwal) [inline src/psim/simulations/single_orbit.cpp:22]: What is the reason for keeping this model around, now that the attitude-orbit model exists?

File: src/psim/truth/transform_direction.cpp (hunk 8 above)
Comment (tanishqaggarwal) [reply src/psim/truth/transform_direction.cpp:62]: What is our naming convention? Does `q_body_eci` mean the rotation from ECI to body, or the other way around?

**Review by tanishqaggarwal — APPROVED** [review]
Hooray!!

File: python/psim/simulation.py (hunk 6 above)
Comment (kylekrol) [reply python/psim/simulation.py:64]: This is used, right now, for the snapshot feature that dumps the current state of the simulation upon termination and not in PTest as far as I'm aware.

File: src/psim/truth/transform_direction.cpp (hunk 8 above)
Comment (kylekrol) [reply src/psim/truth/transform_direction.cpp:62]: `q_a_b` means a  rotation from `b` to `a`.

File: python/psim/plugins/snapshot.py (hunk 10 above)
Comment (kylekrol) [reply python/psim/plugins/snapshot.py:48]: Given that we'd only be tracking truth data here, no I don't foresee this changing too much. Also, the feature is only used to generate new test cases so from the perspective I don't think the extra work is really worth it atm.

File: src/psim/simulations/single_orbit.cpp (hunk 11 above)
Comment (kylekrol) [reply src/psim/simulations/single_orbit.cpp:22]: If I end up running orbit estimator testing and don't want/need all the attitude simulation stuff involved. This is actually exactly why I wanted to shift to this type of architecture!

File: config/parameters/truth/detumble.txt (hunk 9 above)
Comment (kylekrol) [reply config/parameters/truth/detumble.txt:8]: Thanks - and yeah @shihaocao will be keeping track of that for now. #266 will help make changes possible from the PTest side as well.

File: config/parameters/truth/base.txt (hunk 2 above)
Comment (kylekrol) [reply config/parameters/truth/base.txt:23]: The constants are in the existing GNC code. However, in order to have the "truth" constants separate from flight software's set of constants we should keep them in this file separately. It'll allow us to vet the constants going to make sure the system doesn't exhibit vastly different behavior given small perturbations in some of the constants.

It's unfortunate that they're are two sets but there should be a separate set for the truth and flight software implementations.

**Pull request merged**